### PR TITLE
Max churn first draft and test

### PR DESCRIPTION
**Pull request by AlirezaDX**

### CODE REVIEW

**Comment by smcheema** [comment]
Good stuff, let's polish this up and then we'll have max-churn down too. 🎉

**Comment by smcheema** [comment]
Did some light refactoring, should look a lot more integrated now.

@AlirezaDX @irfansharif @musanimu let's align on how we can better capture the `churn` API, lot's of things to keep track of, namely `previousAllocation`, `shouldWeErrorOutToUser = false || true` ... etc, let's go over this PR on Friday.

@AlirezaDX @musanimu one of the tests that we added today is taking `~17s` on my machine, let's see where the bottleneck is and patch if possible, all other tests are on the millisecond level.

File: allocator.go
Comment (irfansharif) [inline allocator.go:231]: These helpers are going to go away, but as an aside -- these are bad names. What's `o`?

File: allocator.go
Comment (irfansharif) [inline allocator.go:180]: Instead of an int, you could use struct{}. Then to check if a key exists, you'd do `_, ok := o[r.id][n.id]`. Btw, if you provide a r.id that doesn't exist (cause it's a new range), you'll have an NPE. I think a better way to do it would be creating a private tuple type; a rangeID+nodeID pair. Something like https://github.com/irfansharif/allocator/blob/96fda6c95716d65eaf4999199abae63b06a9ed49/allocator.go#L41.

File: allocator.go
Comment (irfansharif) [inline allocator.go:233]: Please add comments. It's difficult to follow.

File: allocator.go
Comment (smcheema) [inline allocator.go:261]: Same as above.

File: allocator.go
Comment (irfansharif) [inline allocator.go:237]: Why are we creating new intvars here? Why not re-use the literals we already have for the rangeID-nodeID pair? I think we can use literals + sum over literals like so:

```go
if a.Options.DisableMaxChurn { // Limit the number of items that are moved around.
  if len(a.LastPlacement) != 0 {
    var literals []solver.Literal
    for _, bin := range a.Bins {
      for _, item := range a.Items {
        if !a.placed(item, bin) {
          continue
        }

        // This item-bin pair was previously placed. If it's not
        // placed in the same way this round (the literal will be
        // assigned false). Since we want to count it the number of
        // times this occurs, get its negation.
        literals = append(literals, a.get(item, bin).Not())
      }
    }
    model.AddConstraints(
      solver.NewAtMostKConstraint(a.MaxChurn(), literals...), // Ensure that at most K new placements happen.
    )
    model.Minimize(solver.Sum(literals...)) // Minimize the number of new placements.    
  }
}
```

File: allocator.go
Comment (smcheema) [inline allocator.go:264]: Why the `/2`? I think for full correctness this would be `sum over ranges * saidRanges.rf`.

File: allocator.go
Comment (irfansharif) [inline allocator.go:257]: Delete this.

File: allocator_test.go
Comment (irfansharif) [inline allocator_test.go:186]: General comment: these tests are getting a bit boiler-platey + unreadable. It'd be a good time to try the datadriven format we talked about.

File: allocator_test.go
Comment (irfansharif) [inline allocator_test.go:187]: Add a comment for each test explaining what it's doing.
```go

// TestMaxChurnWithIncreasedNodeCount ...
func TestMaxChurnWithIncreasedNodeCount(t *testing.T) {
```

File: allocator_test.go
Comment (irfansharif) [inline allocator_test.go:188]: Use smaller numbers, and change only a single variable, to make it obvious that the max churn limit should definitely find an answer (or not find an answer). Is testing two variants (increased node count/decreased node count) helpful? I think you should instead simply keep rf=1, nodes=1, and vary number of ranges + max churn limit to be either less than the number of new ranges (infeasible) or greater than it (feasible).

**Review by irfansharif — CHANGES_REQUESTED** [review]

File: allocator_test.go
Comment (smcheema) [inline allocator_test.go:187]: Can you add a couple more tests with smaller `range + node` sets that are easier to assert on explicitly.

File: allocator.go
Comment (irfansharif) [reply allocator.go:227]: Pulling the if conditions outside is more idiomatic.

File: allocator.go
Comment (musanimu) [inline allocator.go:257]: Where are you calling this function? Cant seem to find where its being called

File: allocator.go
Comment (musanimu) [reply allocator.go:125]: Or we could disable max churn option in case of an invalid input and continue, We could just print out a warning or something.

File: allocator.go
Comment (musanimu) [reply allocator.go:227]: I prefer calling it in the function, but since this function is called in only once in the code, Pulling the if statement out to have the same consistency as the rest of the code also makes sense.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:122]: if maxChurn == inf (math.MaxInt64) then it would not apply maxChurn

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:124]: Go's default error messages don't start with uppercase or end with period. I guess following that would be appropriate but I am not sure.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:120]: This is converting the previous result of the solver which is of format `map[RangeID][]NodeID`.
Do we want the result of the solver to be of format `map[RangeID]map[NodeID]int`?

File: allocator.go
Comment (smcheema) [reply allocator.go:124]: Reading Go defined panic messages and that indeed seems to be the case. Maybe it's a ploy to convince developers of the error's severity, i.e whatever you did got me[Go] panicking so bad that I forgot capitalization & punctuation. 😛

File: allocator.go
Comment (smcheema) [reply allocator.go:122]: Let's see what this looks like once we consolidate Max/Min churn together, this one seems like it can be simplified.

File: allocator.go
Comment (irfansharif) [reply allocator.go:124]: The lowercase is indeed the convention; it's because sometimes error messages can be combined upper in the stack and using capitals mid sentence ends up not composing well.

File: allocator.go
Comment (irfansharif) [reply allocator.go:125]: Failing fast seems the most obvious+least surprising thing to do.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:264]: `vars` here are only really added to the list if their value was zero, so there is a good chance that not all node/range values are zeroes so I decided to make the capacity half so that even if the capacity is small, the double of it would be the maximum possible.
(Typically if an append occurs to a list that would run out of capacity would copy the list and allocate double the capacity)

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:229]: Yes... will be removed anyways. Another approach could be applying it to the offset of the linear equation.

File: allocator.go
Comment (smcheema) [reply allocator.go:125]: I think this whole thing can be more simplified if we implicitly try to balance instead of asking callers to have it specified. Having balancing exposed makes sense if they're to pass in some custom comparator/heuristic that we can use to control our spread, but if it's going to be a simplistic approach, might as well do that implicitly and have this be only for `maxChurn`. Thoughts @AlirezaDX @irfansharif @musanimu ?

File: allocator.go
Comment (smcheema) [reply allocator.go:183]: @AlirezaDX this is a valid concern, can we have this addressed?

File: allocator.go
Comment (smcheema) [inline allocator.go:262]: You can declare `toMinimizeTheSumLiterals` as an array of `IntVar`, and then that would eliminate the need for this method, `Literal` and `IntVar` both implement the same interface. Although you'd then need to use `LinearConstraint` as opposed to `NewAtMostKConstraint`.

**Comment by smcheema** [comment]
Looks good overall, just the opts, map, and constraint comments from my side.

File: allocator.go
Comment (musanimu) [reply allocator.go:125]: Agreed if we are not planning to expand on it , A simple approach would be sufficient

**Review by musanimu — APPROVED** [review]
All good from my side.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:125]: Sorry I am not sure I follow, do you mean there is no need to specify max churn?

I think we should keep both around to be able to test how minimizing vs max churn constraint vs both would affect the performance.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:183]: Creating a private tuple would cause a loss of information. Since the map would be only indexed with the combination of rangeID+nodeID, there would be no easy way to understand whether the range is missing or the node if needed.

The nil value of a map is an empty map, therefore the usage below would not throw an NPE given an unknown range:
```GoLang
if previousAssignment, ok := prevAssignmentMap[r.id][n.id]; ok == true && previousAssignment == 1
```

Given the two above reasons I don't think doing so is necessary.

Comment (AlirezaDX) [reply allocator.go:183]: @irfansharif let me know if I am missing the part where an NPE can occur.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:262]: The way it is now might not be as optimized but is more readable don't you think?

Comment (AlirezaDX) [reply allocator.go:262]: By the way, the pattern` Sum(asIntVars(literals)...)` is used multiple times even in the Solver library at `constraint.go`. Perhaps `asIntVars` needs to be a public function.

Any opinion on this @irfansharif

File: allocator.go
Comment (smcheema) [reply allocator.go:183]: https://programming.guide/go/default-zero-value.html, wouldn't the map returned by an invalid `r.id` be `nil`?

File: allocator.go
Comment (smcheema) [reply allocator.go:125]: Actually never-mind, we can't implicitly balance since we need the previous allocation, which we're banking on the caller to have cached. Can we make the difference between minimizing churn and setting a hard limit more explicit by the naming? Also don't think we need to do :

`opt.churnOptions.withMaxChurn = maxChurn < math.MaxInt64`

If the caller asks for maxChurn with that value, why reject if it's at `math.MaxInt64`? If they don't want to use maxChurn with a limit set, they have the option of simply not passing in that option.

File: allocator.go
Comment (smcheema) [reply allocator.go:262]: Yeah both implementations are fine by me. 👍

Comment (irfansharif) [reply allocator.go:262]: It's public: https://godocs.io/github.com/irfansharif/solver#AsIntVars

File: allocator.go
Comment (irfansharif) [reply allocator.go:183]: Yea, what Saad said. Basically it's possible for a new range ID to be passed in during a second call to Allocate, and that would NPE out. I don't understand your loss of information comment. We're not currently checking if a particular node or range exists. Besides, I think all this map stuff goes away once we start modeling replica vars as int vars instead if literals. I'll defer to you guys for what you want to do for this PR -- it'd be nice to merge soon.

**Comment by smcheema** [comment]
@AlirezaDX I think there's only a couple minor things left before we can merge this PR, the opts comment I had above and the map issue. Would it be possible to have this done today? That'll give me some time to integrate all this and migrate it into the `IntVars` prototype.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:125]: Are you suggesting to have to methods that provide the options for Churn? One that takes a `MaxChurn `value and one that does not?

As far as I understood, Irfan suggested that we avoid creating multiple functions that do the same thing and instead similar to the OR tool wrapper utilize values such as `MaxInt` to do the above. I don't mind though to add 2 methods.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:183]: Please give my comments a chance and run some test code:
```Golang
func TestMap(t *testing.T) {
	var myMap = make(map[int]map[int]int)
	myMap[2] = map[int]int{2: 1, 3: 2}

	fmt.Println(myMap)
	fmt.Println(myMap[1][1])
	if _, ok := myMap[1][1]; ok {
		fmt.Println("Not printing this AND no NPE")
	}
}
```
Above wouldn't throw an NPE. Let me know if I am still missing the part that it can be throwing an NPE.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:262]: Not public on the version we are using in this PR, will update to use it accordingly afterwards.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:125]: I believe in another comment Irfan mentioned to use `MaxChurn = 0` as the flag for no `MaxChurn` constraint. I would argue that `MaxInt` is a more natural choice and less likely for the user of the client to use by mistake.

File: allocator.go
Comment (AlirezaDX) [reply allocator.go:183]: Note that a `nil` map is only read as an empty map and writing to the map would actually cause an NPE.

Comment (smcheema) [reply allocator.go:183]: `A nil map is equivalent to an empty map except that no elements may be added.` From : https://golang.org/ref/spec#Map_types
Interesting, I didn't read into it deep enough.

File: allocator.go
Comment (smcheema) [reply allocator.go:262]: I'll fix this once I put my PR up since I bumped up deps, it's fine to merge this as is ATM.

File: allocator.go
Comment (smcheema) [reply allocator.go:125]: I think it's better to not have a flag for maxChurn and simply use the value, we can set it to be -1 initially, and then once someone says `withMaxChurnSet(x)`, we can check if `x >= 0`, and assign it to maxChurn. Then when we want to constrain maxChurn, we check if maxChurn is unset and if so, don't apply the constraint. With maxInt it get's very obscure, i.e currently if I do : 
` allocator.withMaxChurn(maxInt64)`, it's the same as not calling `withMaxChurn` at all. Why give users an option that with a specific argument renders the whole call obsolete? I think we should keep it as simple as possible, have a value for maxChurn initialized to some constant, if a user calls `withMaxChurn`, sanitize the arg, set maxChurn equal to it, and apply the constraint.

Comment (AlirezaDX) [reply allocator.go:125]: What happens if the user only uses `withMaxChurnSet(x)` without providing the previous assignment?

**Review by smcheema — APPROVED** [review]

**Comment by smcheema** [comment]
Synced up with @AlirezaDX, will explore any potential improvements once we integrate with `IntVars` prototype.

**Pull request merged**
